### PR TITLE
net: Add support for DSCP/ECN IPv4/IPv6 header fields

### LINKED ISSUE #50139: net: ipv4: Add DSCP/ToS based QoS support
QoS is implemented at IPv4 using DSCP/ToS fields in the header, these fields are currently hardcoded to 0 in Zephyr's networking stack, neither in TX nor in RX.

E.g., Tx - no option to send ICMP packets with a specificed DSCP
         Rx - if we received an ICMP packet with DSCP set, the reply uses 0 DSCP (In Linux the same DSCP is replicated).

### PR DESCRIPTION
**Pull request by rlubos**
Currently there's no way to specify the value of DSCP/ECN IPv4/IPv6 header fields for outgoing packets, as well as those fields are not parsed from incoming packets. This PR attempts to fix this, by introducing the following changes:

* Specify the DSCP/ECN values at net_pkt level,
* Add processing of those fields in IPv4/IPv6 modules,
* Add new options for net_context and socket, which allow to specify the fields value on a net_context/socket,
* Add support for setting the fields value at ICMP level,
* Finally, introduce new options for `net shell` and `zperf` commands, which allow to configure DSCP/ECN value for outgoing packets.

Resolves #50139

### CODE REVIEW

**Comment by rlubos** [comment]
@krish2718 Please have a look.

File: subsys/net/ip/icmpv6.c [hunk 1]
```diff
@@ -358,6 +362,9 @@ int net_icmpv6_send_echo_request(struct net_if *iface,
 		return -ENOMEM;
 	}
 
+	net_pkt_set_ip_dscp(pkt, (tc & NET_IPV6_DSCP_MASK) >> NET_IPV6_DSCP_OFFSET);
+	net_pkt_set_ip_ecn(pkt, tc & NET_IPV6_ECN_MASK);
```
Comment (jukkar) [inline subsys/net/ip/icmpv6.c:366]: Same thing here

**Review by jukkar — COMMENTED** [review]
Looks good, minor comment about API use.

File: subsys/net/ip/icmpv4.c [hunk 2]
```diff
@@ -533,6 +537,9 @@ int net_icmpv4_send_echo_request(struct net_if *iface,
 		return -ENOMEM;
 	}
 
+	net_pkt_set_ip_dscp(pkt, (tos & NET_IPV4_DSCP_MASK) >> NET_IPV4_DSCP_OFFSET);
+	net_pkt_set_ip_ecn(pkt, tos & NET_IPV4_ECN_MASK);
```
[the remark this answers is not in the document]
Comment (rlubos) [reply subsys/net/ip/icmpv4.c:541]: Sure, that'd be cleaner, I'll provide helper functions.

File: subsys/net/ip/Kconfig [hunk 3]
```diff
@@ -628,6 +628,13 @@ config NET_CONTEXT_SNDBUF
 	  For TCP sockets, the sndbuf will determine the total size of queued
 	  data in the TCP layer.
 
+config NET_CONTEXT_DSCP_ECN
+	bool "Add support for setting DSCP/ECN IP properties on net_context"
+	help
```
Comment (krish2718) [inline subsys/net/ip/Kconfig:633]: IMHO we should make this default

File: subsys/net/ip/net_shell.c [hunk 4]
```diff
@@ -4084,7 +4088,7 @@ static int ping_ipv4(const struct shell *shell,
 #define remove_ipv4_ping_handler()
 #endif /* CONFIG_NET_IPV4 */
 
-static int parse_arg(size_t *i, size_t argc, char *argv[])
+static int parse_arg(size_t *i, size_t argc, char *argv[], bool allow_hex)
```
Comment (krish2718) [inline subsys/net/ip/net_shell.c:4091]: Do we really need this flag? We can allow HEX setting for options, no harm right?

File: subsys/net/ip/Kconfig (hunk 3 above)
Comment (rlubos) [reply subsys/net/ip/Kconfig:633]: Ok, no issues with that on my side.

File: subsys/net/ip/net_shell.c (hunk 4 above)
Comment (rlubos) [reply subsys/net/ip/net_shell.c:4091]: Yeah, on second thought I think you're right.

**Comment by rlubos** [comment]
@jukkar @krish2718 Thanks for the review. I've updated the PR with the following changes:
* Added IPv4/6 getter/setter for DSCP and ECN fields
* Made the net_context (and thus socket) options enabled by default
* dropped the unneeded `allow_hex` flag from `net ping` parser.

**Review by krish2718 — DISMISSED** [review]
LGTM. Nit, IIUC, we are using `TOS` for applications and `DSCP_ECN` in networking stack, this is fine in general as ECN is only used by routers, but we could rename `TOS` to `DSCP_ECN` as well just to be consistent.

**Review by jukkar — DISMISSED** [review]

**Comment by rlubos** [comment]
> Nit, IIUC, we are using TOS for applications and DSCP_ECN in networking stack, this is fine in general as ECN is only used by routers, but we could rename TOS to DSCP_ECN as well just to be consistent.

This is true, there is some variation with the naming, but here's my rationale:
* For the application side (ping/zperf/sockets) I've tried to keep compatibility with Linux nomenclature, to avoid confusion,
* For IPv4/IPv6 specific code, I sticked to the original field names, as the naming is already used in the header
* The generic part of network stack uses `dscp_ecn` as it should work with both IPv4/IPv6.

I hope that make sense.

**Review by tbursztyka — DISMISSED** [review]
Please, make this optional (CONFIG_NET_IP_QOS ?). Not everyone is going to need this (by far), moreover it's adding 1 byte for each net_pkt.

**Comment by rlubos** [comment]
@tbursztyka I've added `CONFIG_NET_IP_DSCP_ECN` to enable DSCP/ECN processing at IP and net_pkt level. I've left the option enabled by default.

**Review by jukkar — DISMISSED** [review]

**Review by krish2718 — APPROVED** [review]

**Comment by rlubos** [comment]
@tbursztyka Can you please revisit?

**Review by krish2718 — DISMISSED** [review]

**Comment by rlubos** [comment]
@tbursztyka Please revisit.

**Review by jukkar — APPROVED** [review]
LGTM

**Review by krish2718 — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**